File: services/hygieneLogic.ts


import { HygieneItem, HygieneCategory } from '../types';

/**
 * HYGIENE MODULE LOGIC CORE
 * Philosophy: "Fighting Entropy"
 */

export enum HygieneStatus {
  FRESH = 'FRESH',   // 0.0 - 0.5
  NORMAL = 'NORMAL', // 0.5 - 1.0
  DUSTY = 'DUSTY',   // 1.0 - 1.5
  MESSY = 'MESSY'    // > 1.5
}

// Default Configuration Map
export const HYGIENE_DEFAULT_CONFIG: Record<HygieneCategory, Omit<HygieneItem, 'id' | 'last_cleaned_at'>> = {
  stove: { category: 'stove', name: '灶台', base_interval_days: 2, is_public_area: true, preferred_unit: 'DAYS' },
  floor_vac: { category: 'floor_vac', name: '地面吸尘', base_interval_days: 3, is_public_area: true, preferred_unit: 'DAYS' },
  toilet: { category: 'toilet', name: '马桶', base_interval_days: 7, is_public_area: true, preferred_unit: 'WEEKS' },
  floor_mop: { category: 'floor_mop', name: '地面拖洗', base_interval_days: 7, is_public_area: true, preferred_unit: 'WEEKS' },
  bedding: { category: 'bedding', name: '床上四件套', base_interval_days: 14, is_public_area: false, preferred_unit: 'WEEKS' },
  washer: { category: 'washer', name: '洗衣机自洁', base_interval_days: 30, is_public_area: true, preferred_unit: 'MONTHS' },
  ac_filter: { category: 'ac_filter', name: '空调滤网', base_interval_days: 90, is_public_area: true, preferred_unit: 'MONTHS' },
  curtain: { category: 'curtain', name: '窗帘', base_interval_days: 180, is_public_area: true, preferred_unit: 'MONTHS' },
};

export const INITIAL_HYGIENE_ITEMS: HygieneItem[] = Object.keys(HYGIENE_DEFAULT_CONFIG).map((key, index) => {
  const config = HYGIENE_DEFAULT_CONFIG[key as HygieneCategory];
  // Stagger initial clean times so they don't all expire at once for the demo
  const staggerDays = index * 2; 
  return {
    id: `h_item_${key}`,
    last_cleaned_at: Date.now() - (staggerDays * 24 * 3600 * 1000), 
    ...config
  };
});

/**
 * Calculates the entropy progress for a specific item.
 * 
 * Formula: Progress = (CurrentTime - LastCleanedTime) / (BaseThreshold / LoadFactor)
 * 
 * @param lastCleanedAt Timestamp (ms)
 * @param baseIntervalDays Days
 * @param isPublicArea Boolean
 * @param householdSize Number of people
 * @param currentTime Timestamp (ms)
 * @returns number (Progress score: 0.0 to 2.0+)
 */
// [PORTABLE]
export const calculate_entropy = (
  lastCleanedAt: number,
  baseIntervalDays: number,
  isPublicArea: boolean,
  householdSize: number,
  currentTime: number
): number => {
  const ONE_DAY_MS = 24 * 3600 * 1000;
  
  // 1. Calculate Load Factor
  // Default 1.0. If > 2 people and public area, 1.2 (dirty 20% faster)
  let loadFactor = 1.0;
  if (householdSize > 2 && isPublicArea) {
    loadFactor = 1.2;
  }

  // 2. Calculate Effective Threshold (ms)
  // Threshold shrinks as load factor increases
  const effectiveThresholdMs = (baseIntervalDays * ONE_DAY_MS) / loadFactor;

  // 3. Calculate Elapsed Time
  const elapsedMs = Math.max(0, currentTime - lastCleanedAt);

  // 4. Calculate Progress
  if (effectiveThresholdMs === 0) return 100; // Safety
  return elapsedMs / effectiveThresholdMs;
};

/**
 * Determines the status enum based on entropy progress.
 */
// [PORTABLE]
export const get_hygiene_status = (progress: number): HygieneStatus => {
  if (progress <= 0.5) return HygieneStatus.FRESH;
  if (progress <= 1.0) return HygieneStatus.NORMAL;
  if (progress <= 1.5) return HygieneStatus.DUSTY;
  return HygieneStatus.MESSY;
};

/**
 * Returns UI-friendly copy based on status.
 */
export const get_status_copy = (status: HygieneStatus, daysSince: number): string => {
  switch (status) {
    case HygieneStatus.FRESH:
      return "干净得在发光 ✨";
    case HygieneStatus.NORMAL:
      return daysSince === 0 ? "刚刚好" : `上次打扫是 ${daysSince} 天前`;
    case HygieneStatus.DUSTY:
      return "好像该照顾一下了 🌿";
    case HygieneStatus.MESSY:
      return "这里需要一次大扫除 🛁";
    default:
      return "";
  }
};

File: Nest/Nest/HygieneLogic.swift
import Foundation

// [PORTABLE] HYGIENE MODULE LOGIC CORE
// Ported from: services/hygieneLogic.ts
// Reference: prd/清洁管理.md

// MARK: - Status Enum

enum HygieneStatus: String, Codable {
    case fresh  = "FRESH"   // 0.0 – 0.5
    case normal = "NORMAL"  // 0.5 – 1.0
    case dusty  = "DUSTY"   // 1.0 – 1.5
    case messy  = "MESSY"   // > 1.5
}

// MARK: - Logic

class HygieneLogic {

    /// [PORTABLE] Calculates the entropy progress (0.0 → ∞) for a hygiene item.
    ///
    /// Formula (PRD §2.1):
    ///   Progress = elapsed / (baseThreshold / LoadFactor)
    ///
    /// LoadFactor (PRD §2.1):
    ///   Default = 1.0
    ///   IF householdSize > 2 AND is_public_area → LoadFactor = 1.2
    ///
    /// NOTE: No random variance — PRD mandates deterministic algorithms.
    ///
    /// Ported from: hygieneLogic.ts → calculate_entropy()
    static func calculateEntropy(
        lastCleanedAt: Date,
        baseIntervalDays: Double,
        isPublicArea: Bool,
        householdMembers: Int = 2,
        currentTime: Date = Date()
    ) -> Double {
        let elapsedHours = currentTime.timeIntervalSince(lastCleanedAt) / 3600.0
        let baseIntervalHours = baseIntervalDays * 24.0

        // PRD: LoadFactor = 1.2 when householdSize > 2 AND public area
        let loadFactor: Double = (isPublicArea && householdMembers > 2) ? 1.2 : 1.0

        let effectiveIntervalHours = baseIntervalHours / loadFactor

        guard effectiveIntervalHours > 0 else { return 0 }
        return max(0.0, elapsedHours / effectiveIntervalHours)
    }

    /// [PORTABLE] Maps entropy value to a status enum.
    ///
    /// Thresholds (PRD §2.2):
    ///   Fresh  0.0 – 0.5
    ///   Normal 0.5 – 1.0
    ///   Dusty  1.0 – 1.5
    ///   Messy  > 1.5
    ///
    /// Ported from: hygieneLogic.ts → get_hygiene_status()
    static func getHygieneStatus(entropy: Double) -> HygieneStatus {
        if entropy < 0.5  { return .fresh }
        if entropy < 1.0  { return .normal }
        if entropy < 1.5  { return .dusty }
        return .messy
    }
    
    /// PRD §4.2 Friendly tone copywriting
    static func getCopy(status: HygieneStatus, itemName: String) -> String {
        switch status {
        case .fresh:  return "干净得在发光 ✨"
        case .normal: return "上次打扫是 \(Int(0)) 天前"   // caller should format days
        case .dusty:  return "好像该照顾一下了 🌱"
        case .messy:  return "焕新一下心情吧 🛁"
        }
    }
}
